Bug: panic on concurrent websocket write

```
{"level":"info","ts":1647711204.5711026,"msg":"session jObTyFWy :: 65b963eb-baac-46f1-9e97-6bf5f2d0682c (172.20.0.6:52232) connected","trace-id":"75e7e3c4-31ba-444a-86f9-aacfc268657d","session-id":"jObTyFWy"}                                                                                       
{"level":"info","ts":1647711204.6305425,"msg":"incoming request (200) GET /b/jObTyFWy/config","trace-id":"c47edc1d-6256-41eb-8453-b3c18a1a878f","session-id":"jObTyFWy","Req.Header.Accept":"application/json, text/plain, */*","Req.Header.Connection":"close","Req.Header.Origin":"http://localhost:3
000","Req.Header.Accept-Encoding":"gzip, deflate, br","Req.Header.X-Real-Ip":"62.216.209.224","Req.Header.User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36","Req.Header.Sec-Ch-Ua-Mobile":"?0","Req.Header.Sec-Ch-
Ua-Platform":"\"macOS\"","Req.Header.Sec-Fetch-Mode":"cors","Resp.Body":"{\"users\":{\"65b963eb-baac-46f1-9e97-6bf5f2d0682c\":{\"id\":\"65b963eb-baac-46f1-9e97-6bf5f2d0682c\",\"alias\":\"YuckyHamster\",\"color\":\"#29AA5B\"}},\"config\":{\"id\":\"jObTyFWy\",\"host\":\"65b963eb-baac-46f1-9e97-6b
f5f2d0682c\",\"maxUsers\":4,\"readOnly\":false}}","Req.Header.Sec-Fetch-Site":"cross-site","Req.Header.Sec-Ch-Ua":"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"99\", \"Google Chrome\";v=\"99\"","Resp.Header.X-Frame-Options":"SAMEORIGIN","Resp.Header.Content-Encoding":"gzip","Req.Header.Accept-La
nguage":"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7","Resp.Header.Vary":"Origin;Accept-Encoding","Req.Header.Referer":"http://localhost:3000/","Resp.Header.X-Xss-Protection":"1; mode=block","Resp.Header.X-Content-Type-Options":"nosniff","Req.Header.Content-Type":"application/json","Req.Header.Sec-Fetc
h-Dest":"empty","Req.Header.X-Forwarded-For":"62.216.209.224","Req.Header.Boardsite-User-Id":"65b963eb-baac-46f1-9e97-6bf5f2d0682c","Resp.Header.Access-Control-Allow-Origin":"*","Resp.Header.Content-Type":"application/json; charset=UTF-8","Req.HttpMethod":"GET","Req.Path":"/b/jObTyFWy/config","
duration":"546.693µs"}                                                   
{"level":"info","ts":1647711204.6924717,"msg":"incoming request (200) GET /b/jObTyFWy/pages/sync","trace-id":"4d421407-53c0-4541-8e76-fa5dcb43b9af","session-id":"jObTyFWy","duration":"1.950085ms","Req.HttpMethod":"GET","Req.Header.X-Forwarded-For":"62.216.209.224","Resp.Header.Content-Encoding"
:"gzip","Req.Header.Sec-Fetch-Site":"cross-site","Req.Header.Connection":"close","Req.Header.Sec-Ch-Ua":"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"99\", \"Google Chrome\";v=\"99\"","Req.Header.Origin":"http://localhost:3000","Resp.Header.Vary":"Origin;Accept-Encoding","Resp.Header.Content-Typ
e":"application/json; charset=UTF-8","Req.Header.Content-Type":"application/json","Req.Header.Sec-Ch-Ua-Mobile":"?0","Req.Header.Accept":"application/json, text/plain, */*","Req.Header.Sec-Ch-Ua-Platform":"\"macOS\"","Req.Header.X-Real-Ip":"62.216.209.224","Req.Header.Accept-Language":"de-DE,de
;q=0.9,en-US;q=0.8,en;q=0.7","Req.Path":"/b/jObTyFWy/pages/sync","Req.Header.Accept-Encoding":"gzip, deflate, br","Req.Header.Boardsite-User-Id":"65b963eb-baac-46f1-9e97-6bf5f2d0682c","Resp.Header.Access-Control-Allow-Origin":"*","Resp.Header.X-Frame-Options":"SAMEORIGIN","Req.Header.Referer":"
http://localhost:3000/","Req.Header.Sec-Fetch-Mode":"cors","Req.Header.Sec-Fetch-Dest":"empty","Resp.Body":"{\"pageRank\":[],\"pages\":{}}","Req.Header.User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36","Resp.He
ader.X-Xss-Protection":"1; mode=block","Resp.Header.X-Content-Type-Options":"nosniff"}                                                             
{"level":"info","ts":1647711204.7692618,"msg":"incoming request (201) POST /b/jObTyFWy/pages","trace-id":"df5c9dd2-229f-4ae5-953a-2e85d1021db5","session-id":"jObTyFWy","Req.Header.User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.484
4.51 Safari/537.36","Req.Header.Accept-Language":"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7","Req.Header.Content-Type":"application/json","Resp.Header.X-Xss-Protection":"1; mode=block","Resp.Header.X-Content-Type-Options":"nosniff","Req.Header.Sec-Fetch-Mode":"cors","Req.Header.Boardsite-User-Id":"65
b963eb-baac-46f1-9e97-6bf5f2d0682c","Req.Header.Sec-Ch-Ua":"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"99\", \"Google Chrome\";v=\"99\"","Resp.Header.Access-Control-Allow-Origin":"*","Req.Path":"/b/jObTyFWy/pages","Req.Header.Sec-Ch-Ua-Platform":"\"macOS\"","Req.Body":"{\"pageId\":[\"f8ah1Yft\
"],\"index\":[0],\"meta\":{\"f8ah1Yft\":{\"size\":{\"width\":620,\"height\":877},\"background\":{\"style\":\"blank\"}}},\"strokes\":{\"f8ah1Yft\":{}}}","duration":"2.722467ms","Req.HttpMethod":"POST","Req.Header.Sec-Fetch-Dest":"empty","Resp.Header.Content-Encoding":"gzip","Req.Header.Content-L
ength":"148","Req.Header.X-Real-Ip":"62.216.209.224","Resp.Header.X-Frame-Options":"SAMEORIGIN","Resp.Header.Vary":"Origin;Accept-Encoding","Req.Header.Origin":"http://localhost:3000","Req.Header.Connection":"close","Req.Header.Referer":"http://localhost:3000/","Req.Header.Sec-Ch-Ua-Mobile":"?0
","Req.Header.Sec-Fetch-Site":"cross-site","Req.Header.Accept":"application/json, text/plain, */*","Req.Header.X-Forwarded-For":"62.216.209.224","Req.Header.Accept-Encoding":"gzip, deflate, br"}
{"level":"info","ts":1647711210.7728288,"msg":"Close session jObTyFWy","trace-id":"75e7e3c4-31ba-444a-86f9-aacfc268657d","session-id":"jObTyFWy"}
{"level":"info","ts":1647711210.7728975,"msg":"session jObTyFWy :: 65b963eb-baac-46f1-9e97-6bf5f2d0682c (172.20.0.6:52232) disconnected","trace-id":"75e7e3c4-31ba-444a-86f9-aacfc268657d","session-id":"jObTyFWy"}
{"level":"info","ts":1647711210.7731845,"msg":"incoming request (200) GET /b/jObTyFWy/users/65b963eb-baac-46f1-9e97-6bf5f2d0682c/socket","trace-id":"75e7e3c4-31ba-444a-86f9-aacfc268657d","session-id":"jObTyFWy","Req.Header.X-Forwarded-For":"62.216.209.224","Req.Header.Upgrade":"websocket","Resp.Header.Vary":"Origin;Accept-Encoding","Resp.Header.X-Frame-Options":"SAMEORIGIN","Req.Header.Accept-Encoding":"gzip, deflate, br","Req.Header.Sec-Websocket-Key":"47fDyrLpC1SoPUKyFNhpag==","Req.Header.Connection":"upgrade","Req.HttpMethod":"GET","Req.Header.X-Real-Ip":"62.216.209.224","Req.Header.User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36","Resp.Header.Access-Control-Allow-Origin":"*","duration":"6.2024526s","Req.Header.Sec-Websocket-Version":"13","Resp.Header.X-Content-Type-Options":"nosniff","Req.Header.Accept-Language":"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7","Req.Header.Sec-Websocket-Extensions":"permessage-deflate; client_max_window_bits","Req.Header.Origin":"http://localhost:3000","Resp.Header.Content-Encoding":"gzip","Resp.Header.X-Xss-Protection":"1; mode=block","Req.Header.Pragma":"no-cache","Req.Header.Cache-Control":"no-cache","Req.Path":"/b/jObTyFWy/users/65b963eb-baac-46f1-9e97-6bf5f2d0682c/socket"}
{"level":"info","ts":1647712834.6580229,"msg":"session O20299Ur :: 747f2387-ebaa-4312-b198-25e277cc395d (172.20.0.6:52138) disconnected","trace-id":"3f200d16-2fae-44ac-bffd-b213d5fec2f1","session-id":"O20299Ur"}
panic: concurrent write to websocket connection

goroutine 7614 [running]:
github.com/gorilla/websocket.(*messageWriter).flushFrame(0x4000551170, 0x1, {0x0, 0x0, 0x0})
        /go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:610 +0x508
github.com/gorilla/websocket.(*messageWriter).Close(0x4000551170)
        /go/pkg/mod/github.com/gorilla/websocket@v1.4.2/conn.go:724 +0x58
github.com/gorilla/websocket.(*Conn).WriteJSON(0x4000298580, {0x59dbe0, 0x400037cfc0})
        /go/pkg/mod/github.com/gorilla/websocket@v1.4.2/json.go:29 +0xe4
github.com/heat1q/boardsite/session.(*broadcaster).broadcastLoop(0x40003a46e0)
        /app/session/broadcast.go:102 +0x5cc
created by github.com/heat1q/boardsite/session.(*broadcaster).Bind
        /app/session/broadcast.go:59 +0x74
```

--- a/api/websocket/websocket.go
+++ b/api/websocket/websocket.go
@@ -38,7 +38,6 @@ func onClientConnect(ctx context.Context, scb session.Controller, userID string,
 func onClientDisconnect(ctx context.Context, scb session.Controller, userID string, conn *gws.Conn) {
 	scb.UserDisconnect(ctx, userID)
 	log.Ctx(ctx).Infof("session %s :: %s (%s) disconnected", scb.ID(), userID, conn.RemoteAddr().String())
-	closeWS(conn, nil)
 }
 
 // Subscribe subscribes to the websocket connection
@@ -50,7 +49,8 @@ func Subscribe(c echo.Context, scb session.Controller, userID string) error {
 	}
 
 	if err := onClientConnect(ctx, scb, userID, conn); err != nil {
-		closeWS(conn, gws.FormatCloseMessage(gws.CloseNormalClosure, fmt.Sprintf("%v", err)))
+		_ = conn.WriteMessage(gws.CloseMessage, gws.FormatCloseMessage(gws.CloseNormalClosure, fmt.Sprintf("%v", err)))
+		_ = conn.Close()
 		return err
 	}
 	defer onClientDisconnect(ctx, scb, userID, conn)
@@ -77,9 +77,4 @@ func Subscribe(c echo.Context, scb session.Controller, userID string) error {
 		}
 	}
 	return nil
-}
-
-func closeWS(conn *gws.Conn, message []byte) {
-	_ = conn.WriteMessage(gws.CloseMessage, message)
-	_ = conn.Close()
 }

--- a/session/broadcast.go
+++ b/session/broadcast.go
@@ -2,7 +2,9 @@ package session
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"runtime"
 
 	gws "github.com/gorilla/websocket"
 
@@ -10,6 +12,8 @@ import (
 	"github.com/heat1q/boardsite/api/types"
 	"github.com/heat1q/boardsite/redis"
 )
+
+var ErrBroadcasterClosed = errors.New("broadcaster: closed")
 
 //counterfeiter:generate . Broadcaster
 type Broadcaster interface {
@@ -94,38 +98,55 @@ func (b *broadcaster) getUsers() map[string]*User {
 func (b *broadcaster) broadcastLoop() {
 	for {
 		users := b.getUsers()
-		select {
-		case data := <-b.broadcast:
-			for userID, user := range users { // Send to all connected clients
-				// except the origin, i.e. the initiator of message
-				if userID != data.Sender {
-					if err := user.Conn.WriteJSON(data); err != nil {
-						log.Global().Warnf("cannot broadcast to %s: %v",
-							user.Conn.RemoteAddr(), err)
-					}
-				}
-			}
-		case data := <-b.send:
-			u, ok := users[data.Receiver]
-			if !ok {
-				log.Global().Warnf("broadcastLoop: unknown receiver %v", data.Receiver)
-				continue
-			}
-			if err := u.Conn.WriteJSON(data); err != nil {
-				log.Global().Warnf("broadcastLoop: %v", err)
-			}
-		case data := <-b.control:
-			u, ok := users[data.Receiver]
-			if !ok {
-				log.Global().Warnf("broadcastLoop: unknown receiver %v", data.Receiver)
-				continue
-			}
-			msg := gws.FormatCloseMessage(gws.CloseNormalClosure, fmt.Sprintf("%v", data.Content))
-			_ = u.Conn.WriteMessage(gws.CloseMessage, msg)
-		case <-b.close:
+		err := b.broadcastToUser(users)
+		if errors.Is(err, ErrBroadcasterClosed) {
 			return
 		}
+		if err != nil {
+			log.Global().Warnf("broadcastLoop: %v", err)
+		}
 	}
+}
+
+func (b *broadcaster) broadcastToUser(users map[string]*User) error {
+	defer func() {
+		if r := recover(); r != nil {
+			stack := make([]byte, 2<<12)
+			length := runtime.Stack(stack, true)
+			log.Global().Errorf("[PANIC RECOVER] %v %s", r, stack[:length])
+		}
+	}()
+
+	select {
+	case data := <-b.broadcast:
+		for userID, user := range users { // Send to all connected clients
+			// except the origin, i.e. the initiator of message
+			if userID != data.Sender {
+				if err := user.Conn.WriteJSON(data); err != nil {
+					log.Global().Warnf("cannot broadcast to %s: %v",
+						user.Conn.RemoteAddr(), err)
+				}
+			}
+		}
+	case data := <-b.send:
+		u, ok := users[data.Receiver]
+		if !ok {
+			return fmt.Errorf("send: unkown receiver: %v", data.Receiver)
+		}
+		if err := u.Conn.WriteJSON(data); err != nil {
+			return fmt.Errorf("send: writeJSON: %w", err)
+		}
+	case data := <-b.control:
+		u, ok := users[data.Receiver]
+		if !ok {
+			return fmt.Errorf("control: unkown receiver: %v", data.Receiver)
+		}
+		msg := gws.FormatCloseMessage(gws.CloseNormalClosure, fmt.Sprintf("%v", data.Content))
+		_ = u.Conn.WriteMessage(gws.CloseMessage, msg)
+	case <-b.close:
+		return ErrBroadcasterClosed
+	}
+	return nil
 }
 
 // dbUpdateLoop updates database according to given Stroke values

--- a/session/user.go
+++ b/session/user.go
@@ -164,9 +164,11 @@ func (scb *controlBlock) UserConnect(userID string, conn *gws.Conn) error {
 // Broadcast that user has disconnected from session.
 func (scb *controlBlock) UserDisconnect(ctx context.Context, userID string) {
 	scb.muUsr.Lock()
-	u := scb.users[userID]
-	delete(scb.users, u.ID)
-	scb.numUsers--
+	u, ok := scb.users[userID]
+	if ok {
+		delete(scb.users, u.ID)
+		scb.numUsers--
+	}
 	numCl := scb.numUsers
 	scb.muUsr.Unlock()
 
@@ -181,6 +183,11 @@ func (scb *controlBlock) UserDisconnect(ctx context.Context, userID string) {
 	scb.broadcaster.Broadcast() <- types.Message{
 		Type:    MessageTypeUserDisconnected,
 		Content: u,
+	}
+
+	scb.broadcaster.Control() <- types.Message{
+		Receiver: userID,
+		Content:  "Closed by server",
 	}
 }
 

